Little kernel bootloader support

alfonsosanchezbeato opened the pull request:
Add little kernel bootloader support.

sil2100 commented on ubuntu_image/tests/test_builder.py:
```diff
@@ -768,6 +771,152 @@ def test_populate_filesystems_bogus_type(self):
                 str(cm.exception), 'Invalid part filesystem type: 801')
 
+    def test_populate_filesystems_lk_bootloader(self):
+        # We do a bit-wise copy when the file system has no type.
```
This comment should describe the actual test that's being performed. I know it's been copy-pasted a few times before, but let's try making it better and getting back to actually documenting the test cases better.

sil2100 commented on ubuntu_image/tests/test_builder.py:
```diff
@@ -768,6 +771,152 @@ def test_populate_filesystems_bogus_type(self):
+                (part0_img, part0_path, 'snap', 'hold the door'))
+
+    def test_populate_filesystems_lk_bootloader_no_boot_files(self):
+        # We do a bit-wise copy when the file system has no type.
```
Same here.

sil2100 commented on ubuntu_image/common_builder.py:
```diff
@@ -349,6 +349,23 @@ def prepare_filesystems(self):
+        # one comes from the kernel snap, while the second one is modified by
+        # 'snap prepare-image' to set the right core and kernel for the kernel
+        # command line.
+        if volume.bootloader is BootLoader.lk:
```
A few things here. I'm wondering if this is the best fit for the boot copies to gadget actually. We do something similar-ish _populate_one_bootfs() - since basically this is part of preparation of the bootfs, right? Is there any reason why we do it in _populate_one_volume instead (which is more for population of non-boot partitions)?

The other thing is that I guess at first I was a bit concerned about this being run per-volume. Since we run this for every volume defined in the gadget, so in theory, conceptually, all the copies should be done per-volume. But on the other hand, I doubt we'll actually have any devices that would have 2+ volumes from which 2 have BootLoader.lk specified - there doesn't seem to be any practical reasons for that. So I guess it's all good from this POV.

alfonsosanchezbeato replied:
I do not think we can do this in `_populate_one_bootfs()`, as what is done there is populating a partition with files. In this case we need image files that are directly the partition content. And those files do not come from the gadget:
* boot.img comes from the kernel snap - we need it there so kernel updates can be performed
* snapbootsel.bin is modified by 'snap prepare', similarly to what is done to uboot/grub environment files - we need to take it after that has happened

Re: per-volume, I agree, I do not think we would have a case with more than one volume with lk.

sil2100 replied:
Ah! Ok, that makes much more sense, this wasn't entirely clear to me without seeing a gadget.yaml - now looking at the one in the test-case below, I finally understand how this works. My original understanding was that it was files that had to be present on the filesystem for those partitions.

So in this case I think that conceptually this change basically doesn't really fit in either - so let's leave it as is. It could be a potential good fit for _populate_one_bootfs() because it's basically 'preparing' boot-related files from the gadget for future copying over into the image, but it also fits _populate_one_volume() because that's where the binary copying is happening. Anyway, it's good as is, no need to bikeshed!

sil2100 commented:
A few other things that I forgot to mention in my previous review: First, I'm thinking if maybe we should add some sanity-checking logic to the parser code to, for instance, make sure the bootimg and bootselect roles are only used on volumes that have the lk bootloader defined in the gadget.yaml.

Other thing I'd like clarification on: since in ubuntu-image we don't use the two new roles, I assume they're used by snap prepare-image somehow, right?

alfonsosanchezbeato commented:
Thanks for the review! I have updated the PR now. Some comments:
* I have updated https://forum.snapcraft.io/t/the-gadget-snap/696 too. I have asked the snapd team to review, but note that the new partition roles are already in snapd anyway: https://github.com/snapcore/snapd/blob/master/gadget/gadget.go#L116
* Yes, `snap prepare-image` uses the new roles, the 'bootselect' partition contains the data needed for kernel / core reverts (equivalent to try_core, try_kernel, etc,), while 'bootimg' is used to identify partitions containing kernel images.
* Right, it might make sense to do some additional checking in parser.py - I will take a look at that

alfonsosanchezbeato commented:
@sil2100 I have changed now the partition names, following https://github.com/snapcore/snapd/pull/7572
